Random issues discovered while testing for test_continuous_random_attach.py

### Your Environment

- **Version:** Local Development Setup
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** Magma MME/SPGW service
- **Deployment Environment:** Vagrant (AGW)

### Setup Details

The S1APTester ([https://github.com/facebookexperimental/S1APTester](https://github.com/facebookexperimental/S1APTester)) was used along with Magma EPC in Vagrant based setup to reproduce the issue.

### Describe the Issue
The test case **test_continuous_random_attach.py** shows unstable behavior while executed multiple times and with number of UEs changed. The different issues observed are as follows:

- MME crash for stale ENB context
While running the test case in loop with a local script for 100 UEs, the MME was found to be crashing while handling S1 Setup after few iterations of the test case execution. After analysis the issue was observed in the stale ENB cleanup function when handling the s1 setup. The ENB state from hash tables was tried to be cleaned-up twice from the function, once indirectly from s1ap_remove_enb and another directly from s1ap_remove_ue. The backtrace from the syslog for the crash is as follows: (syslog file attached :syslog_test_continuous_random_attach_issue1). 

  - **To Reproduce**
For reproducing the issue run the test case in loop with runner script (https://drive.google.com/drive/folders/1D4jyi0GW82xgdObiCvkAXwol4kxjosdJ?usp=sharing)

```
Mar 31 10:24:17 magma-dev mme[18877]: =================================================================
Mar 31 10:24:17 magma-dev mme[18877]: ==18877==ERROR: AddressSanitizer: heap-use-after-free on address 0x617000263380 at pc 0x55cfebbeef50 bp 0x7f6fc419d3f0 sp 0x7f6fc419d3e8
Mar 31 10:24:17 magma-dev mme[18877]: WRITE of size 4 at 0x617000263380 thread T24
Mar 31 10:24:17 magma-dev mme[18877]:     #0 0x55cfebbeef4f in s1ap_remove_enb /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:576
Mar 31 10:24:17 magma-dev mme[18877]:     #1 0x55cfebc7bb94 in clean_stale_enb_state /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:371
Mar 31 10:24:17 magma-dev mme[18877]:     #2 0x55cfebc811e2 in s1ap_mme_handle_s1_setup_request /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:613
Mar 31 10:24:17 magma-dev mme[18877]:     #3 0x55cfebc7a631 in s1ap_mme_handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:225
Mar 31 10:24:17 magma-dev mme[18877]:     #4 0x55cfebbeac32 in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:137
Mar 31 10:24:17 magma-dev mme[18877]:     #5 0x7f6fdb6407bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Mar 31 10:24:17 magma-dev mme[18877]:     #6 0x55cfebbec07b in s1ap_mme_thread /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:320
Mar 31 10:24:17 magma-dev mme[18877]:     #7 0x7f6fdde1f4a3 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x74a3)
Mar 31 10:24:17 magma-dev mme[18877]:     #8 0x7f6fd947bd0e in __clone (/lib/x86_64-linux-gnu/libc.so.6+0xe8d0e)
Mar 31 10:24:17 magma-dev mme[18877]: 0x617000263380 is located 0 bytes inside of 648-byte region [0x617000263380,0x617000263608)
Mar 31 10:24:17 magma-dev mme[18877]: freed by thread T24 here:
Mar 31 10:24:17 magma-dev mme[18877]:     #0 0x7f6fde302a10 in free (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc1a10)
Mar 31 10:24:17 magma-dev mme[18877]:     #1 0x55cfebb87c1a in free_wrapper /home/vagrant/magma/lte/gateway/c/oai/common/dynamic_memory_check.c:47
Mar 31 10:24:17 magma-dev mme[18877]:     #2 0x55cfebbe69dd in hashtable_ts_free /home/vagrant/magma/lte/gateway/c/oai/lib/hashtable/hashtable.c:850
Mar 31 10:24:17 magma-dev mme[18877]:     #3 0x55cfebbef004 in s1ap_remove_enb /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:578
Mar 31 10:24:17 magma-dev mme[18877]:     #4 0x55cfebbeee22 in s1ap_remove_ue /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:566
Mar 31 10:24:17 magma-dev mme[18877]:     #5 0x55cfebc7b7af in clean_stale_enb_state /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:365
Mar 31 10:24:17 magma-dev mme[18877]:     #6 0x55cfebc811e2 in s1ap_mme_handle_s1_setup_request /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:613
Mar 31 10:24:17 magma-dev mme[18877]:     #7 0x55cfebc7a631 in s1ap_mme_handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:225
Mar 31 10:24:17 magma-dev mme[18877]:     #8 0x55cfebbeac32 in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:137
Mar 31 10:24:17 magma-dev mme[18877]:     #9 0x7f6fdb6407bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Mar 31 10:24:17 magma-dev mme[18877]: previously allocated by thread T24 here:
Mar 31 10:24:17 magma-dev mme[18877]:     #0 0x7f6fde302ed0 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.3+0xc1ed0)
Mar 31 10:24:17 magma-dev mme[18877]:     #1 0x55cfebbecd82 in s1ap_new_enb /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:455
Mar 31 10:24:17 magma-dev mme[18877]:     #2 0x55cfebc9cf4e in s1ap_handle_new_association /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c:2370
Mar 31 10:24:17 magma-dev mme[18877]:     #3 0x55cfebbeb05a in handle_message /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:161
Mar 31 10:24:17 magma-dev mme[18877]:     #4 0x7f6fdb6407bd in zloop_start (/usr/lib/x86_64-linux-gnu/libczmq.so.4+0x287bd)
Mar 31 10:24:17 magma-dev mme[18877]: Thread T24 created by T0 here:
Mar 31 10:24:17 magma-dev mme[18877]:     #0 0x7f6fde271f59 in __interceptor_pthread_create (/usr/lib/x86_64-linux-gnu/libasan.so.3+0x30f59)
Mar 31 10:24:17 magma-dev mme[18877]:     #1 0x55cfecae6f57 in itti_create_task /home/vagrant/magma/lte/gateway/c/oai/lib/itti/intertask_interface.c:322
Mar 31 10:24:17 magma-dev mme[18877]:     #2 0x55cfebbec48d in s1ap_mme_init /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:345
Mar 31 10:24:17 magma-dev mme[18877]:     #3 0x55cfebb8061e in main /home/vagrant/magma/lte/gateway/c/oai/oai_mme/oai_mme.c:137
Mar 31 10:24:17 magma-dev mme[18877]:     #4 0x7f6fd93b32e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
Mar 31 10:24:17 magma-dev mme[18877]: SUMMARY: AddressSanitizer: heap-use-after-free /home/vagrant/magma/lte/gateway/c/oai/tasks/s1ap/s1ap_mme.c:576 in s1ap_remove_enb
Mar 31 10:24:17 magma-dev mme[18877]: Shadow bytes around the buggy address:
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044620: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044630: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044640: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044650: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044660: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]: =>0x0c2e80044670:[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044680: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e80044690: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e800446a0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e800446b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Mar 31 10:24:17 magma-dev mme[18877]:   0x0c2e800446c0: fd fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Mar 31 10:24:17 magma-dev mme[18877]: Shadow byte legend (one shadow byte represents 8 application bytes):
Mar 31 10:24:17 magma-dev mme[18877]:   Addressable:           00
Mar 31 10:24:17 magma-dev mme[18877]:   Partially addressable: 01 02 03 04 05 06 07
Mar 31 10:24:17 magma-dev mme[18877]:   Heap left redzone:       fa
Mar 31 10:24:17 magma-dev mme[18877]:   Heap right redzone:      fb
Mar 31 10:24:17 magma-dev mme[18877]:   Freed heap region:       fd
Mar 31 10:24:17 magma-dev mme[18877]:   Stack left redzone:      f1
Mar 31 10:24:17 magma-dev mme[18877]:   Stack mid redzone:       f2
Mar 31 10:24:17 magma-dev mme[18877]:   Stack right redzone:     f3
Mar 31 10:24:17 magma-dev mme[18877]:   Stack partial redzone:   f4
Mar 31 10:24:17 magma-dev mme[18877]:   Stack after return:      f5
Mar 31 10:24:17 magma-dev mme[18877]:   Stack use after scope:   f8
Mar 31 10:24:17 magma-dev mme[18877]:   Global redzone:          f9
Mar 31 10:24:17 magma-dev mme[18877]:   Global init order:       f6
Mar 31 10:24:17 magma-dev mme[18877]:   Poisoned by user:        f7
Mar 31 10:24:17 magma-dev mme[18877]:   Container overflow:      fc
Mar 31 10:24:17 magma-dev mme[18877]:   Array cookie:            ac
Mar 31 10:24:17 magma-dev mme[18877]:   Intra object redzone:    bb
Mar 31 10:24:17 magma-dev mme[18877]:   ASan internal:           fe
Mar 31 10:24:17 magma-dev mme[18877]:   Left alloca redzone:     ca
Mar 31 10:24:17 magma-dev mme[18877]:   Right alloca redzone:    cb
Mar 31 10:24:17 magma-dev mme[18877]: ==18877==ABORTING
Mar 31 10:24:17 magma-dev systemd[1]: magma@mme.service: Main process exited, code=exited, status=1/FAILURE
Mar 31 10:24:18 magma-dev systemd[1]: magma@mme.service: Unit entered failed state.
Mar 31 10:24:18 magma-dev systemd[1]: magma@mme.service: Failed with result 'exit-code'.
```
[syslog_test_continuous_random_attach_issue1.txt](https://github.com/magma/magma/files/6233895/syslog_test_continuous_random_attach_issue1.txt)

- S1APTester crashes for memory corruption with python backtrace and memory chunk while handling the UE count of 240 UEs for same test case "test_continuous_random_attach.py". The issue is not always reproducible and occurs randomly sometimes.

```
*** Error in `/home/vagrant/build/python/bin/python3.5': malloc(): memory corruption: 0x00007fda844bf400 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x70bfb)[0x7fda9e77dbfb]
/lib/x86_64-linux-gnu/libc.so.6(+0x76fc6)[0x7fda9e783fc6]
/lib/x86_64-linux-gnu/libc.so.6(+0x79089)[0x7fda9e786089]
/lib/x86_64-linux-gnu/libc.so.6(__libc_malloc+0x54)[0x7fda9e787f64]
/home/vagrant/s1ap-tester/bin/libtfw.so(+0x304148)[0x7fda9825f148]
/home/vagrant/s1ap-tester/bin/libtfw.so(SAlloc+0x87)[0x7fda982600ad]
/home/vagrant/s1ap-tester/bin/libtfw.so(SGetDBuf+0x8f)[0x7fda98256cb3]
/home/vagrant/s1ap-tester/bin/libtfw.so(SAddPreMsgMult+0x15c)[0x7fda98255533]
/home/vagrant/s1ap-tester/bin/libtfw.so(SPkU32+0x59)[0x7fda98259508]
/home/vagrant/s1ap-tester/bin/libtfw.so(cmPkUeDelReq+0x8f)[0x7fda98046c27]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbIfmDamUeDelReq+0x2e)[0x7fda980461ee]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbBuildAndSendS1ShutdownReq+0x224)[0x7fda9804b2aa]
/home/vagrant/s1ap-tester/bin/libtfw.so(NbUiNbtMsgReq+0x23a)[0x7fda98060efc]
/home/vagrant/s1ap-tester/bin/libtfw.so(cmUnPkNbtMsgReq+0x91)[0x7fda980f4d6d]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbActvTsk+0xd5)[0x7fda98045c66]
/home/vagrant/s1ap-tester/bin/libtfw.so(+0x2f7223)[0x7fda98252223]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x74a4)[0x7fda9f6024a4]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7fda9e7f5d0f]
======= Memory map: ========
55f6c976b000-55f6c9b5b000 r-xp 00000000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
55f6c9d5b000-55f6c9d5d000 r--p 003f0000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
55f6c9d5d000-55f6c9df4000 rw-p 003f2000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
55f6c9df4000-55f6c9e25000 rw-p 00000000 00:00 0
55f6c9ff6000-55f6cb089000 rw-p 00000000 00:00 0                          [heap]
7fda6c000000-7fda6c024000 rw-p 00000000 00:00 0
7fda6c024000-7fda70000000 ---p 00000000 00:00 0
7fda706b9000-7fda707f9000 rw-p 00000000 00:00 0
7fda74000000-7fda74127000 rw-p 00000000 00:00 0
7fda74127000-7fda78000000 ---p 00000000 00:00 0
7fda78000000-7fda7806f000 rw-p 00000000 00:00 0
7fda7806f000-7fda7c000000 ---p 00000000 00:00 0
7fda7c000000-7fda7e1ea000 rw-p 00000000 00:00 0
7fda7e1ea000-7fda80000000 ---p 00000000 00:00 0
7fda80000000-7fda8002b000 rw-p 00000000 00:00 0
7fda8002b000-7fda84000000 ---p 00000000 00:00 0
7fda84000000-7fda84702000 rw-p 00000000 00:00 0
7fda84702000-7fda88000000 ---p 00000000 00:00 0
7fda88000000-7fda88975000 rw-p 00000000 00:00 0
7fda88975000-7fda8c000000 ---p 00000000 00:00 0
7fda8c000000-7fda8c029000 rw-p 00000000 00:00 0
7fda8c029000-7fda90000000 ---p 00000000 00:00 0
7fda9000b000-7fda9004b000 rw-p 00000000 00:00 0
7fda9004b000-7fda9006a000 r-xp 00000000 08:01 2119712                    /usr/lib/x86_64-linux-gnu/libiperf.so.0.0.0
7fda9006a000-7fda9026a000 ---p 0001f000 08:01 2119712                    /usr/lib/x86_64-linux-gnu/libiperf.so.0.0.0
7fda9026a000-7fda9026b000 r--p 0001f000 08:01 2119712                    /usr/lib/x86_64-linux-gnu/libiperf.so.0.0.0
7fda9026b000-7fda9026c000 rw-p 00020000 08:01 2119712                    /usr/lib/x86_64-linux-gnu/libiperf.so.0.0.0
7fda9026c000-7fda9026e000 r-xp 00000000 08:01 661104                     /home/vagrant/s1ap-tester/bin/libtrfgen.so
7fda9026e000-7fda9046d000 ---p 00002000 08:01 661104                     /home/vagrant/s1ap-tester/bin/libtrfgen.so
7fda9046d000-7fda9046e000 r--p 00001000 08:01 661104                     /home/vagrant/s1ap-tester/bin/libtrfgen.so
7fda9046e000-7fda9046f000 rw-p 00002000 08:01 661104                     /home/vagrant/s1ap-tester/bin/libtrfgen.so
7fda9046f000-7fda90470000 ---p 00000000 00:00 0
7fda90470000-7fda90c70000 rw-p 00000000 00:00 0
7fda90c70000-7fda90c71000 ---p 00000000 00:00 0
7fda90c71000-7fda91471000 rw-p 00000000 00:00 0
7fda91471000-7fda91472000 ---p 00000000 00:00 0
7fda91472000-7fda91c72000 rw-p 00000000 00:00 0
7fda91c72000-7fda91c73000 ---p 00000000 00:00 0
7fda91c73000-7fda92473000 rw-p 00000000 00:00 0
7fda92473000-7fda92474000 ---p 00000000 00:00 0
7fda92474000-7fda92c74000 rw-p 00000000 00:00 0
7fda92c74000-7fda92c75000 ---p 00000000 00:00 0
7fda92c75000-7fda93475000 rw-p 00000000 00:00 0
7fda93475000-7fda93855000 rw-s 00000000 00:08 36129                      socket:[36129]
7fda93855000-7fda93856000 ---p 00000000 00:00 0
7fda93856000-7fda94056000 rw-p 00000000 00:00 0
7fda94056000-7fda94057000 ---p 00000000 00:00 0
7fda94057000-7fda94857000 rw-p 00000000 00:00 0
7fda94857000-7fda94858000 ---p 00000000 00:00 0
7fda94858000-7fda95058000 rw-p 00000000 00:00 0
7fda95058000-7fda95059000 ---p 00000000 00:00 0
7fda95059000-7fda95859000 rw-p 00000000 00:00 0
7fda95859000-7fda9585a000 ---p 00000000 00:00 0
7fda9585a000-7fda9605a000 rw-p 00000000 00:00 0
7fda9605a000-7fda9605b000 ---p 00000000 00:00 0
7fda9605b000-7fda9685b000 rw-p 00000000 00:00 0
7fda9685b000-7fda9685c000 ---p 00000000 00:00 0
7fda9685c000-7fda9705c000 rw-p 00000000 00:00 0
7fda9705c000-7fda9705d000 ---p 00000000 00:00 0
7fda9705d000-7fda97f5b000 rw-p 00000000 00:00 0
7fda97f5b000-7fda982da000 r-xp 00000000 08:01 661103                     /home/vagrant/s1ap-tester/bin/libtfw.so
7fda982da000-7fda984d9000 ---p 0037f000 08:01 661103                     /home/vagrant/s1ap-tester/bin/libtfw.so
7fda984d9000-7fda984dc000 r--p 0037e000 08:01 661103                     /home/vagrant/s1ap-tester/bin/libtfw.so
7fda984dc000-7fda985a1000 rw-p 00381000 08:01 661103                     /home/vagrant/s1ap-tester/bin/libtfw.so
7fda985a1000-7fda98aa4000 rw-p 00000000 00:00 0
7fda98aa4000-7fda98aca000 r-xp 00000000 08:01 2228277                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7fda98aca000-7fda98cc9000 ---p 00026000 08:01 2228277                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7fda98cc9000-7fda98ccd000 r--p 00025000 08:01 2228277                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7fda98ccd000-7fda98cce000 rw-p 00029000 08:01 2228277                    /lib/x86_64-linux-gnu/libtinfo.so.5.9
7fda98cce000-7fda98d12000 r-xp 00000000 08:01 2229756                    /lib/x86_64-linux-gnu/libreadline.so.7.0
7fda98d12000-7fda98f11000 ---p 00044000 08:01 2229756                    /lib/x86_64-linux-gnu/libreadline.so.7.0
7fda98f11000-7fda98f13000 r--p 00043000 08:01 2229756                    /lib/x86_64-linux-gnu/libreadline.so.7.0
7fda98f13000-7fda98f19000 rw-p 00045000 08:01 2229756                    /lib/x86_64-linux-gnu/libreadline.so.7.0
7fda98f19000-7fda98f1b000 rw-p 00000000 00:00 0
7fda98f1b000-7fda98f21000 r-xp 00000000 08:01 133119                     /usr/lib/python3.5/lib-dynload/readline.cpython-35m-x86_64-linux-gnu.so
7fda98f21000-7fda99120000 ---p 00006000 08:01 133119                     /usr/lib/python3.5/lib-dynload/readline.cpython-35m-x86_64-linux-gnu.so
7fda99120000-7fda99121000 r--p 00005000 08:01 133119                     /usr/lib/python3.5/lib-dynload/readline.cpython-35m-x86_64-linux-gnu.so
7fda99121000-7fda99123000 rw-p 00006000 08:01 133119                     /usr/lib/python3.5/lib-dynload/readline.cpython-35m-x86_64-linux-gnu.so
7fda99123000-7fda99127000 r-xp 00000000 08:01 133121                     /usr/lib/python3.5/lib-dynload/termios.cpython-35m-x86_64-linux-gnu.so
7fda99127000-7fda99326000 ---p 00004000 08:01 133121                     /usr/lib/python3.5/lib-dynload/termios.cpython-35m-x86_64-linux-gnu.so
7fda99326000-7fda99327000 r--p 00003000 08:01 133121                     /usr/lib/python3.5/lib-dynload/termios.cpython-35m-x86_64-linux-gnu.so
7fda99327000-7fda99329000 rw-p 00004000 08:01 133121                     /usr/lib/python3.5/lib-dynload/termios.cpython-35m-x86_64-linux-gnu.so
7fda99329000-7fda9932d000 r-xp 00000000 08:01 2228327                    /lib/x86_64-linux-gnu/libuuid.so.1.3.0
7fda9932d000-7fda9952c000 ---p 00004000 08:01 2228327                    /lib/x86_64-linux-gnu/libuuid.so.1.3.0
7fda9952c000-7fda9952d000 r--p 00003000 08:01 2228327                    /lib/x86_64-linux-gnu/libuuid.so.1.3.0
7fda9952d000-7fda9952e000 rw-p 00004000 08:01 2228327                    /lib/x86_64-linux-gnu/libuuid.so.1.3.0
7fda9952e000-7fda99631000 r-xp 00000000 08:01 2100413                    /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
7fda99631000-7fda99830000 ---p 00103000 08:01 2100413                    /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
7fda99830000-7fda99833000 r--p 00102000 08:01 2100413                    /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
7fda99833000-7fda99835000 rw-p 00105000 08:01 2100413                    /usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
7fda99835000-7fda99836000 rw-p 00000000 00:00 0
7fda99836000-7fda9984a000 r-xp 00000000 08:01 133107                     /usr/lib/python3.5/lib-dynload/_sqlite3.cpython-35m-x86_64-linux-gnu.so
7fda9984a000-7fda99a4a000 ---p 00014000 08:01 133107                     /usr/lib/python3.5/lib-dynload/_sqlite3.cpython-35m-x86_64-linux-gnu.so
7fda99a4a000-7fda99a4b000 r--p 00014000 08:01 133107                     /usr/lib/python3.5/lib-dynload/_sqlite3.cpython-35m-x86_64-linux-gnu.so
7fda99a4b000-7fda99a4e000 rw-p 00015000 08:01 133107                     /usr/lib/python3.5/lib-dynload/_sqlite3.cpython-35m-x86_64-linux-gnu.so
7fda99a4e000-7fda99e0f000 rw-p 00000000 00:00 0
7fda99e10000-7fda99e1a000 r-xp 00000000 08:01 1447753                    /home/vagrant/build/python/lib/python3.5/site-packages/simplejson/_speedups.cpython-35m-x86_64-linux-gnu.so
7fda99e1a000-7fda9a019000 ---p 0000a000 08:01 1447753                    /home/vagrant/build/python/lib/python3.5/site-packages/simplejson/_speedups.cpython-35m-x86_64-linux-gnu.so
7fda9a019000-7fda9a01a000 r--p 00009000 08:01 1447753                    /home/vagrant/build/python/lib/python3.5/site-packages/simplejson/_speedups.cpython-35m-x86_64-linux-gnu.so
7fda9a01a000-7fda9a01b000 rw-p 0000a000 08:01 1447753                    /home/vagrant/build/python/lib/python3.5/site-packages/simplejson/_speedups.cpython-35m-x86_64-linux-gnu.so
7fda9a01b000-7fda9a053000 r-xp 00000000 08:01 2105542                    /usr/lib/x86_64-linux-gnu/libmpdec.so.2.4.2
7fda9a053000-7fda9a252000 ---p 00038000 08:01 2105542                    /usr/lib/x86_64-linux-gnu/libmpdec.so.2.4.2
7fda9a252000-7fda9a253000 r--p 00037000 08:01 2105542                    /usr/lib/x86_64-linux-gnu/libmpdec.so.2.4.2
7fda9a253000-7fda9a254000 rw-p 00038000 08:01 2105542                    /usr/lib/x86_64-linux-gnu/libmpdec.so.2.4.2
7fda9a254000-7fda9a278000 r-xp 00000000 08:01 133101                     /usr/lib/python3.5/lib-dynload/_decimal.cpython-35m-x86_64-linux-gnu.so
7fda9a278000-7fda9a477000 ---p 00024000 08:01 133101                     /usr/lib/python3.5/lib-dynload/_decimal.cpython-35m-x86_64-linux-gnu.so
7fda9a477000-7fda9a478000 r--p 00023000 08:01 133101                     /usr/lib/python3.5/lib-dynload/_decimal.cpython-35m-x86_64-linux-gnu.so
7fda9a478000-7fda9a481000 rw-p 00024000 08:01 133101                     /usr/lib/python3.5/lib-dynload/_decimal.cpython-35m-x86_64-linux-gnu.so
7fda9a481000-7fda9a48e000 r-xp 00000000 08:01 133102                     /usr/lib/python3.5/lib-dynload/_json.cpython-35m-x86_64-linux-gnu.so
7fda9a48e000-7fda9a68d000 ---p 0000d000 08:01 133102                     /usr/lib/python3.5/lib-dynload/_json.cpython-35m-x86_64-linux-gnu.so
7fda9a68d000-7fda9a68e000 r--p 0000c000 08:01 133102                     /usr/lib/python3.5/lib-dynload/_json.cpython-35m-x86_64-linux-gnu.so
7fda9a68e000-7fda9a68f000 rw-p 0000d000 08:01 133102                     /usr/lib/python3.5/lib-dynload/_json.cpython-35m-x86_64-linux-gnu.so
7fda9a68f000-7fda9ab0f000 rw-p 00000000 00:00 0
7fda9ab0f000-7fda9ad5a000 r-xp 00000000 08:01 529953                     /usr/local/lib/python3.5/dist-packages/google/protobuf/pyext/_message.cpython-35m-x86_64-linux-gnu.so
7fda9ad5a000-7fda9af59000 ---p 0024b000 08:01 529953                     /usr/local/lib/python3.5/dist-packages/google/protobuf/pyext/_message.cpython-35m-x86_64-linux-gnu.so
7fda9af59000-7fda9af6a000 rw-p 0024a000 08:01 529953                     /usr/local/lib/python3.5/dist-packages/google/protobuf/pyext/_message.cpython-35m-x86_64-linux-gnu.so
7fda9af6a000-7fda9af6d000 rw-p 00000000 00:00 0
7fda9af6d000-7fda9af6e000 r-xp 00000000 08:01 274721                     /usr/local/lib/python3.5/dist-packages/google/protobuf/internal/_api_implementation.cpython-35m-x86_64-linux-gnu.so
7fda9af6e000-7fda9b16d000 ---p 00001000 08:01 274721                     /usr/local/lib/python3.5/dist-packages/google/protobuf/internal/_api_implementation.cpython-35m-x86_64-linux-gnu.so
7fda9b16d000-7fda9b16e000 rw-p 00000000 08:01 274721                     /usr/local/lib/python3.5/dist-packages/google/protobuf/internal/_api_implementation.cpython-35m-x86_64-linux-gnu.so
7fda9b16e000-7fda9b2ee000 rw-p 00000000 00:00 0
7fda9b2ee000-7fda9b2f1000 r-xp 00000000 08:01 133106                     /usr/lib/python3.5/lib-dynload/_multiprocessing.cpython-35m-x86_64-linux-gnu.so
7fda9b2f1000-7fda9b4f0000 ---p 00003000 08:01 133106                     /usr/lib/python3.5/lib-dynload/_multiprocessing.cpython-35m-x86_64-linux-gnu.so
7fda9b4f0000-7fda9b4f1000 r--p 00002000 08:01 133106                     /usr/lib/python3.5/lib-dynload/_multiprocessing.cpython-35m-x86_64-linux-gnu.so
7fda9b4f1000-7fda9b4f2000 rw-p 00003000 08:01 133106                     /usr/lib/python3.5/lib-dynload/_multiprocessing.cpython-35m-x86_64-linux-gnu.so
7fda9b4f2000-7fda9b508000 r-xp 00000000 08:01 2233011                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fda9b508000-7fda9b707000 ---p 00016000 08:01 2233011                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fda9b707000-7fda9b708000 r--p 00015000 08:01 2233011                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fda9b708000-7fda9b709000 rw-p 00016000 08:01 2233011                    /lib/x86_64-linux-gnu/libgcc_s.so.1
7fda9b709000-7fda9b87b000 r-xp 00000000 08:01 2097161                    /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22Aborted
Makefile:33: recipe for target 'integ_test' failed
make: *** [integ_test] Interrupt
```

- S1APTester crashes for memory corruption with python backtrace and memory chunk while handling the UE count of 240 UEs for same test case "test_continuous_random_attach.py". The issue is not always reproducible and occurs randomly sometimes.

```
*** Error in `/home/vagrant/build/python/bin/python3.5': malloc(): memory corruption: 0x00007f057c435230 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x70bfb)[0x7f0593b53bfb]
/lib/x86_64-linux-gnu/libc.so.6(+0x76fc6)[0x7f0593b59fc6]
/lib/x86_64-linux-gnu/libc.so.6(+0x79089)[0x7f0593b5c089]
/lib/x86_64-linux-gnu/libc.so.6(__libc_malloc+0x54)[0x7f0593b5df64]
/home/vagrant/s1ap-tester/bin/libtfw.so(+0x30412d)[0x7f058d5a812d]
/home/vagrant/s1ap-tester/bin/libtfw.so(SAlloc+0x87)[0x7f058d5a9092]
/home/vagrant/s1ap-tester/bin/libtfw.so(SGetDBuf+0x8f)[0x7f058d59fc98]
/home/vagrant/s1ap-tester/bin/libtfw.so(SAddPreMsgMult+0x15c)[0x7f058d59e518]
/home/vagrant/s1ap-tester/bin/libtfw.so(cmPkPtr+0xa7)[0x7f058d58ef5c]
/home/vagrant/s1ap-tester/bin/libtfw.so(cmPkNbuS1RelInd+0x76)[0x7f058d43e976]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbSendS1RelIndToUeApp+0x72)[0x7f058d3a9796]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbBuildAndSendS1ShutdownReq+0x198)[0x7f058d39421e]
/home/vagrant/s1ap-tester/bin/libtfw.so(NbUiNbtMsgReq+0x23a)[0x7f058d3a9ee1]
/home/vagrant/s1ap-tester/bin/libtfw.so(cmUnPkNbtMsgReq+0x91)[0x7f058d43dd52]
/home/vagrant/s1ap-tester/bin/libtfw.so(nbActvTsk+0xd5)[0x7f058d38ec66]
/home/vagrant/s1ap-tester/bin/libtfw.so(+0x2f7208)[0x7f058d59b208]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x74a4)[0x7f05949d84a4]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x3f)[0x7f0593bcbd0f]
======= Memory map: ========
5593eb820000-5593ebc10000 r-xp 00000000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
5593ebe10000-5593ebe12000 r--p 003f0000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
5593ebe12000-5593ebea9000 rw-p 003f2000 08:01 661240                     /home/vagrant/build/python/bin/python3.5
5593ebea9000-5593ebeda000 rw-p 00000000 00:00 0
5593eca5b000-5593edb1c000 rw-p 00000000 00:00 0                          [heap]
7f04ff73f000-7f04ff740000 ---p 00000000 00:00 0
7f04ff740000-7f04fff40000 rw-p 00000000 00:00 0
```

- Sometimes the testscript was found to be running in an infinite loop. But it is printing only the attaching and detaching log with ue id and no other message communication prints were found like authentication or identity etc. As per logs, the sending of S1AP/SCTP messages are failing and in the sctpd some timeouts are also observed while sending messages. The sctp communication was also missing in the packet capture.

- Sometimes S1APTester is crashing with segmentation fault while handling the initial UE message randomly for some UE. The crash occurs while finding the UE in the hashlist (mostly while handling the attach request/initial UE message). The comparision in the hashlist was found to be done over some invalid address.

```
Thread 7 "nosetests" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fe15efd8700 (LWP 1869)]
0x00007fe1621ca748 in cmHashListFind (hashListCp=0x7fe162529040 <nbCb+3296>, key=0x7fe14a1e7320 "\223", keyLen=4, seqNmb=0, entry=0x7fe15efd7dd0)
    at ../src/ssi/cm_hash.c:1283
1283          if (cmHashMatchKey(hashListEnt->key, hashListEnt->keyLen, key, keyLen)
(gdb) bt
#0  0x00007fe1621ca748 in cmHashListFind (hashListCp=0x7fe162529040 <nbCb+3296>, key=0x7fe14a1e7320 "\223", keyLen=4, seqNmb=0, entry=0x7fe15efd7dd0)
    at ../src/ssi/cm_hash.c:1283
#1  0x00007fe161fe0016 in NbHandleInitialUeMsg (initialUeMsg=0x7fe14a1e7320) at ../src/enbApp/nb_ue_msg_hndlr.c:70
#2  0x00007fe161fe25aa in NbUiNbuHdlInitialUeMsg (pst=0x7fe15efd7ef0, req=0x7fe14a1e7320) at ../src/enbApp/nb_ui.c:286
#3  0x00007fe162076069 in cmUnPkNbuInitialUeMsg (func=0x7fe161fe2588 <NbUiNbuHdlInitialUeMsg>, pst=0x7fe15efd7ef0, mBuf=0x7fe14a1de5e0)
    at ../src/fw_cm/nbu.c:445
#4  0x00007fe161fc6cbd in nbActvTsk (pst=0x7fe15efd7ef0, mBuf=0x7fe14a1de5e0) at ../src/enbApp/nb_ex_ms.c:127
#5  0x00007fe1621d3223 in mtTskHdlr (tskPtr=0x7fe162a1bff8 <osCp+6296>) at ../src/ssi/mt_ss.c:3472
#6  0x00007fe1695834a4 in start_thread (arg=0x7fe15efd8700) at pthread_create.c:456
#7  0x00007fe168776d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
(gdb) p hashListEnt->key
Cannot access memory at address 0xc0a83c8e00000810
(gdb) p *hashListEnt->key
Cannot access memory at address 0xc0a83c8e00000810
(gdb) p *key
$1 = 147 '\223'
(gdb) p key
$2 = (U8 *) 0x7fe14a1e7320 "\223"
(gdb) p entry
$3 = (U64 *) 0x7fe15efd7dd0
(gdb) p* entry
$4 = 0
(gdb) p initialUeMsg
No symbol "initialUeMsg" in current context.
```

```
Thread 7 "nosetests" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f7a3ad03700 (LWP 4147)]
0x00007f7a3def572d in cmHashListFind (hashListCp=0x7f7a3e254040 <nbCb+3296>, key=0x7f7a3ad02908 <incomplete sequence \313>, keyLen=4, seqNmb=0,
    entry=0x7f7a3ad02900) at ../src/ssi/cm_hash.c:1283
1283          if (cmHashMatchKey(hashListEnt->key, hashListEnt->keyLen, key, keyLen)
(gdb) bt
#0  0x00007f7a3def572d in cmHashListFind (hashListCp=0x7f7a3e254040 <nbCb+3296>, key=0x7f7a3ad02908 <incomplete sequence \313>, keyLen=4, seqNmb=0,
    entry=0x7f7a3ad02900) at ../src/ssi/cm_hash.c:1283
#1  0x00007f7a3dcfb9e1 in nbPrcS1apConCfm (conCfm=0x7f7a3ad02a50) at ../src/enbApp/nb_s1ap_hndlr.c:2469
#2  0x00007f7a3dcf3faf in NbIfmS1apConCfm (pst=0x7f7a2c09ef94, suId=0, conCfm=0x7f7a3ad02a50) at ../src/enbApp/nb_ifm_s1ap.c:602
#3  0x00007f7a3dcf3e01 in UzLiSztConCfm (pst=0x7f7a2c09ef94, suId=0, conCfm=0x7f7a3ad02a50) at ../src/enbApp/nb_ifm_s1ap.c:493
#4  0x00007f7a3de09cc6 in SzUiSztConCfm (pst=0x7f7a2c09ef94, suId=0, conCfm=0x7f7a3ad02a50) at ../src/s1ap/sz_ptui.c:686
#5  0x00007f7a3de10c55 in szUiSndConCfm (conCb=0x7f7a2c2a5300, conCfm=0x7f7a3ad02a50) at ../src/s1ap/sz_ui.c:2222
#6  0x00007f7a3de0b0a9 in szSmRcvConnStMsg (conCb=0x7f7a2c2a5300, peerCb=0x7f7a2c11f1a0, pdu=0x7f7a2c534c08) at ../src/s1ap/sz_sm.c:938
#7  0x00007f7a3de0bb5b in szSmExcMt (conCb=0x7f7a2c2a5300, peerCb=0x7f7a2c11f1a0, pdu=0x7f7a2c534c08, evnt=17 '\021', dir=2 '\002')
    at ../src/s1ap/sz_sm.c:1576
#8  0x00007f7a3ddfc54b in szLiCOMsgHandler (peerCb=0x7f7a2c11f1a0, pdu=0x7f7a2c534c08) at ../src/s1ap/sz_li.c:682
#9  0x00007f7a3ddfc238 in szLiRcvFromLower (mBuf=0x0, peerCb=0x7f7a2c11f1a0) at ../src/s1ap/sz_li.c:531
#10 0x00007f7a3ddfe4bf in SzLiSctDatInd (pst=0x7f7a3ad02ef0, suId=0, suAssocId=0, strmId=1, indType=0x7f7a3ad02e88, protId=18, mBuf=0x7f7a2801cc90)
    at ../src/s1ap/sz_li.c:1992
#11 0x00007f7a3dda549a in cmUnpkSctDatInd (func=0x7f7a3ddfe15f <SzLiSctDatInd>, pst=0x7f7a3ad02ef0, mBuf=0x7f7a2801cc90) at ../src/cm/sct.c:3335
#12 0x00007f7a3ddfb5fa in szActvTsk (pst=0x7f7a3ad02ef0, mBuf=0x7f7a2801cc90) at ../src/s1ap/sz_ex_ms.c:299
#13 0x00007f7a3defe208 in mtTskHdlr (tskPtr=0x7f7a3e746ff8 <osCp+6296>) at ../src/ssi/mt_ss.c:3472
#14 0x00007f7a452ae4a4 in start_thread (arg=0x7f7a3ad03700) at pthread_create.c:456
#15 0x00007f7a444a1d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
(gdb) p key
$1 = (U8 *) 0x7f7a3ad02908 <incomplete sequence \313>
(gdb) p *key
$2 = 203 '\313'
(gdb) p keyLen
$3 = 4
(gdb) p hashListEnt->keyLen
Cannot access memory at address 0x18
(gdb) p *hashListEnt->keyLen
Cannot access memory at address 0x18
(gdb) p *hashListEnt->key
Cannot access memory at address 0x10
```

--- a/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c
+++ b/lte/gateway/c/oai/tasks/s1ap/s1ap_mme_handlers.c
@@ -362,13 +362,20 @@ void clean_stale_enb_state(
     ue_description_t* ue_ref = NULL;
     for (int i = 0; i < keys->num_keys; i++) {
       ue_ref = s1ap_state_get_ue_mmeid((mme_ue_s1ap_id_t) keys->keys[i]);
+      /* The function s1ap_remove_ue will take care of removing the enb also,
+       * when the last UE is removed
+       */
       s1ap_remove_ue(state, ue_ref);
     }
     FREE_HASHTABLE_KEY_ARRAY(keys);
+  } else {
+    // Remove the old eNB association
+    OAILOG_INFO(
+        LOG_S1AP, "Deleting eNB: %s (Sctp_assoc_id = %u)",
+        stale_enb_association->enb_name, stale_enb_association->sctp_assoc_id);
+    s1ap_remove_enb(state, stale_enb_association);
   }
 
-  // Remove the old eNB association
-  s1ap_remove_enb(state, stale_enb_association);
   OAILOG_DEBUG(LOG_S1AP, "Removed stale eNB and all associated UEs.");
 }
 

